modules/dnn/src/layers/proposal_layer.cpp

```diff
@@ -148,11 +148,86 @@ class ProposalLayerImpl : public ProposalLayer
         deltasPermute->finalize(layerInputs, layerOutputs);
     }
 
+#ifdef HAVE_OPENCL
+    bool forward_ocl(InputArrayOfArrays inps, OutputArrayOfArrays outs, OutputArrayOfArrays intnals)
+    {
+        std::vector<UMat> inputs;
+        std::vector<UMat> outputs;
+        std::vector<UMat> internals;
+
+        inps.getUMatVector(inputs);
+        outs.getUMatVector(outputs);
+        intnals.getUMatVector(internals);
+
+        CV_Assert(inputs.size() == 3);
+        CV_Assert(internals.size() == 3);
+        const UMat& scores = inputs[0];
+        const UMat& bboxDeltas = inputs[1];
+        const UMat& imInfo = inputs[2];
+        UMat& priorBoxes = internals[0];
+        UMat& permuttedScores = internals[1];
+        UMat& permuttedDeltas = internals[2];
+
+        CV_Assert(imInfo.total() >= 2);
+        // We've chosen the smallest data type because we need just a shape from it.
+        float rows = imInfo.getMat(ACCESS_READ).at<float>(0);
+        float cols = imInfo.getMat(ACCESS_READ).at<float>(1);
```

Comment:
I think more efficient way is to select ROI (assume 2D layout - or use reshape()) and to copy it once (instead of two maps).

```
Mat szMat;
imInfo(Rect(0, 0, 1, 2)).copyTo(szMat);
float rows = (int)szMat.at<float>(0);
float cols = (int)szMat.at<float>(1);
```

Reply:
ok, replaced two maps with one copyTo